filebeat/input/log/harvester.go

```diff
@@ -391,12 +391,12 @@ func (h *Harvester) SendStateUpdate() {
 		return
 	}
 
-	logp.Debug("harvester", "Update state: %s, offset: %v", h.state.Source, h.state.Offset)
-	h.states.Update(h.state)
-
 	d := util.NewData()
 	d.SetState(h.state)
 	h.publishState(d)
+
+	logp.Debug("harvester", "Update state: %s, offset: %v", h.state.Source, h.state.Offset)
+	h.states.Update(h.state)
```

Comment:
Can you elaborate on this move? I would assume the harvester itself should be the first one to know about state updates?

Reply:
Will add my findings to the PR description.

The Update here is no to the harvester, but to the prospector. Thanks to shutdown racing with scan phase we sometimes send finish and remove events out of order if `publishState` is not called before `states.Update`.